modules/apps/27-interchain-accounts/controller/keeper/handshake.go

```diff
@@ -136,3 +137,42 @@ func (Keeper) OnChanCloseConfirm(
 ) error {
 	return nil
 }
+
+// OnChanUpgradeInit performs the upgrade init step of the channel upgrade handshake.
+func (k Keeper) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, connectionHops []string, version string) (string, error) {
+	if strings.TrimSpace(version) == "" {
+		return "", errorsmod.Wrap(icatypes.ErrInvalidVersion, "version cannot be empty")
+	}
+
+	metadata, err := icatypes.ParseMedataFromString(version)
+	if err != nil {
+		return "", err
+	}
+
+	channel, found := k.channelKeeper.GetChannel(ctx, portID, channelID)
+	if !found {
+		return "", errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "failed to retrieve channel %s on port %s", channelID, portID)
+	}
+
+	currentMetadata, err := icatypes.ParseMedataFromString(channel.Version)
+	if err != nil {
+		return "", err
+	}
+
+	if err := icatypes.ValidateControllerMetadata(ctx, k.channelKeeper, connectionHops, metadata); err != nil {
+		return "", errorsmod.Wrap(err, "invalid metadata")
+	}
+
+	// the interchain account address on the host chain
+	// must remain the same after the upgrade.
+	if currentMetadata.Address != metadata.Address {
+		return "", errorsmod.Wrap(icatypes.ErrInvalidAccountAddress, "address cannot be changed")
+	}
+
+	if currentMetadata.ControllerConnectionId != connectionHops[0] {
+		return "", errorsmod.Wrap(connectiontypes.ErrInvalidConnectionIdentifier, "proposed connection hop must not change")
+	}
+
+	metadataBz := icatypes.ModuleCdc.MustMarshalJSON(&metadata)
```

Comment:
I realize now (and this something that could be changed as well in the spec) that `metadata` doesn't change in this function, so we could simply just return `version` and save the marshalling.

Reply:
nice catch!